### PR TITLE
Shooting a BSA beam at the SM will make it delaminate

### PR DESCRIPTION
**Pull request by MrPerson**
## About The Pull Request

Adds a BSA beam blocker element that blocks bluespace artillery beams. SM and singulo have it.
The SM will also call explode() when shot, so maybe don't do that.
Keep in mind these are talking about the beam the artillery launches that goes to the edge of the map that destroys everything in its path. After it hits the edge, it's supposed to explode on what you aimed it at (a gps, normally). It didn't do that before, now it does. That explosion is a regular old explosion, nothing special happens if you blow stuff up with that.

## Why It's Good For The Game

Some chucklefucks on Sybil shot the SM with BSA yesterday and were disappointed nothing happened. Well here's your something.

## Changelog
:cl: MrPerson
add: If you shoot the supermatter crystal with bluespace artillery, you're gonna have a bad time.
fix: Bluespace artillery will blow its targets up properly instead of doing nothing.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
shooting the singularity with the BSA should delete the singularity

makes it a valid way to stop a rampaging singularity.

File: code/modules/station_goals/bsa.dm [hunk 1]
```diff
@@ -178,17 +178,42 @@
...
+		if(SEND_SIGNAL(tile, COMSIG_ATOM_BSA_BEAM) & COMSIG_ATOM_BLOCKS_BSA_BEAM)
+			blocker = tile
+		else
+			for(var/AM in tile)
+				var/atom/movable/stuff = AM
+				if(SEND_SIGNAL(stuff, COMSIG_ATOM_BSA_BEAM) & COMSIG_ATOM_BLOCKS_BSA_BEAM)
+					blocker = stuff
+					break
+		if(blocker) //BYOND's only mechanism for escaping from an inner foreach loop is GOTO. If I did that, this code would GOTO the trash can
+			target = locate(targ_x, targ_y, z) //both get_turf(blocker) and T may be deleted by this point
```
Comment (ninjanomnom) [inline code/modules/station_goals/bsa.dm]: Turf references don't null and just refer to the new turf

File: code/datums/elements/bsa_blocker.dm [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//blocks bluespace artillery beams that try to fly through
+//look not all elements need to be fancy
+/datum/element/bsa_blocker/Attach(datum/target)
+	if(!ismovableatom(target))
```
Comment (ninjanomnom) [inline code/datums/elements/bsa_blocker.dm]: Can't it just be any /atom?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by PepperPrepper** [comment]
> shooting the singularity with the BSA should delete the singularity. makes it a valid way to stop a rampaging singularity.

How about a 50% chance to kill the singularity but a 50% chance to make it go up a stage? That way it's a gamble and not just a certain fix.

**Comment by MrPerson** [comment]
If you somehow manage to actually hit the fucking thing while it's moving, the last thing I would want to do is punish you for the effort.

**Comment by PepperPrepper** [comment]
singularities move fairly slowly. Plus surely it would suck the shot in?

**Comment by MrPerson** [comment]
Sucking the shot in is why I just blocked it originally. Oranges wants it to kill the singulo for gameplay reasons, which is fair enough and frankly gameplay considerations trump everything else.

If it winds up being overpowered and common (somehow???) then sure, make it a 50/50. Until then, no need to punish for something that's already really hard to do.

**Comment by PepperPrepper** [comment]
Yeah fair enough It would be a lot more difficult to pull off then I originally thought. I guess my opinion just comes from my general hatred to guaranteed hard counters in games.

File: code/modules/station_goals/bsa.dm (hunk 1 above)
Comment (MrPerson) [reply code/modules/station_goals/bsa.dm]: I always forget how turfs worked, I'm pretty sure I've literally said this to other people before.

**Comment by necromanceranne** [comment]
What happens if I shoot a supermatter shard?

**Pull request merged**